gitlab/__init__.py

```diff
@@ -342,8 +347,14 @@ def list(self, obj_class, **kwargs):
                 if key in cls_kwargs:
                     del cls_kwargs[key]
 
-            return [cls(self, item, **cls_kwargs) for item in r.json()
-                    if item is not None]
+            results = [cls(self, item, **cls_kwargs) for item in r.json()
+                       if item is not None]
+            if 'next' in r.links and 'url' in r.links['next']:
+                args = kwargs.copy()
+                args['next_url'] = r.links['next']['url']
+                logger.debug("Iterating results 'next' link: %s", args['next_url'])
+                results.extend(self.list(obj_class, **args))
```

Comment:
I'm not sure I see this as "a problem" - the whole reason why I opened this PR is because I have about 200 users, and will soon have about 300 repositories. I've seen some latency when fetching all that, but not much.

I like @dserodio's idea of using a generator for this. I'm going to have a try at implementing it, though I haven't done much with generators before, and perhaps worse, I've never really worked with ``unittest`` or ``httmock``.